airbyte-integrations/postgres-destination/src/main/java/io/airbyte/integrations/destination/postgres/PostgresDestination.java

```diff
@@ -0,0 +1,343 @@
+
+    private static final long GRACEFUL_SHUTDOWN_MINUTES = 5L;
+    private static final int MIN_RECORDS = 500;
+    private static final int BATCH_SIZE = 500;
+
+    private final ScheduledExecutorService writerPool;
+    private final BasicDataSource connectionPool;
+    private final Map<String, WriteConfig> writeConfigs;
+    private final Schema schema;
+
+    public RecordConsumer(BasicDataSource connectionPool, Map<String, WriteConfig> writeConfigs, Schema schema) {
+      this.connectionPool = connectionPool;
+      this.writeConfigs = writeConfigs;
+      this.schema = schema;
+      this.writerPool = Executors.newSingleThreadScheduledExecutor();
+      // todo (cgardens) - how long? boh.
+      Runtime.getRuntime().addShutdownHook(new GracefulShutdownHandler(GRACEFUL_SHUTDOWN_MINUTES, TimeUnit.MINUTES, writerPool));
+
+      writerPool.scheduleWithFixedDelay(
+          () -> writeStreamsWithNRecords(MIN_RECORDS, BATCH_SIZE, writeConfigs, connectionPool),
+          THREAD_DELAY_MILLIS,
+          THREAD_DELAY_MILLIS,
+          TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Write records from buffer to postgres in batch.
+     *
+     * @param minRecords - the minimum number of records in the buffer before writing. helps avoid
+     *        wastefully writing one record at a time.
+     * @param batchSize - the maximum number of records to write in a single insert.
+     * @param writeBuffers - map of stream name to its respective buffer.
+     * @param connectionPool - connection to the db.
+     */
+    private static void writeStreamsWithNRecords(
+                                                 int minRecords,
+                                                 int batchSize,
+                                                 Map<String, WriteConfig> writeBuffers,
+                                                 BasicDataSource connectionPool) {
+      for (final Map.Entry<String, WriteConfig> entry : writeBuffers.entrySet()) {
+        final String tmpTableName = entry.getValue().getTmpTableName();
+        final CloseableInputQueue<byte[]> writeBuffer = entry.getValue().getWriteBuffer();
+        while (writeBuffer.size() > minRecords) {
+          try {
+            DatabaseHelper.query(connectionPool, ctx -> ctx.execute(buildWriteQuery(batchSize, writeBuffer, tmpTableName)));
+          } catch (SQLException e) {
+            throw new RuntimeException(e);
+          }
+        }
+      }
+    }
+
+    // build the following query:
+    // INSERT INTO <tableName>(data)
+    // VALUES
+    // ({ "my": "data" }),
+    // ({ "my": "data" });
+    private static String buildWriteQuery(int batchSize, CloseableInputQueue<byte[]> writeBuffer, String tmpTableName) {
+      final StringBuilder query = new StringBuilder(String.format("INSERT INTO %s(%s)\n", tmpTableName, COLUMN_NAME))
+          .append("VALUES \n");
+      boolean firstRecordInQuery = true;
+      for (int i = 0; i <= batchSize; i++) {
```

Comment:
Do you have an index error here? If it is expected, you should add a comment to explain it since it not a normal code structure

Reply:
made it standard.